File: 2025/148. Sort List/Solution2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode half = split(head);
        ListNode left = sortList(head);
        ListNode right = sortList(half);
        return merge(left, right);
    }

    private ListNode split(ListNode head) {
        ListNode fast = head, slow = head;
        while (fast.next != null && fast.next.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode half = slow.next;
        slow.next = null;
        return half;
    }

    private ListNode merge(ListNode left, ListNode right) {
        ListNode dummy = new ListNode(), tmp = dummy;
        while (left != null && right != null) {
            if (left.val < right.val) {
                tmp.next = left;
                left = left.next;
            } else {
                tmp.next = right;
                right = right.next;
            }
            tmp = tmp.next;
        }
        if (left == null)
            tmp.next = right;
        else
            tmp.next = left;
        return dummy.next;
    }

}

File: 2025/143. Reorder List/solution2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # slow fast find middle
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # reverse half list
        prev, slow = slow, slow.next
        while slow and slow.next: 
            tmp, tmp2 = prev.next, slow.next.next
            prev.next = slow.next
            slow.next.next = tmp
            slow.next = tmp2

        # merge two half lists
        slow, fast = head, prev.next
        prev.next = None
        while slow and fast:
            tmp, tmp2 = slow.next, fast.next
            slow.next = fast
            fast.next = tmp
            slow = fast.next
            fast = tmp2